File: backend/app/schemas/package.py
"""
套餐相关的 Pydantic Schemas
"""
from datetime import datetime
from typing import Optional
from decimal import Decimal
from pydantic import BaseModel, Field


# ========== 创建套餐 ==========
class PackageCreate(BaseModel):
    """创建套餐 Schema"""
    name: str = Field(..., min_length=1, max_length=100, description="套餐名称")
    description: Optional[str] = Field(None, description="套餐描述")

    # 权限配置
    max_accounts: int = Field(10, ge=1, description="最大账号数")
    max_share_pages: int = Field(5, ge=1, description="最大分享页数")
    max_nodes: int = Field(1, ge=1, description="最大节点数")
    min_unlock_interval: int = Field(3600, ge=60, description="最短解锁间隔(秒)")

    # 功能权限
    allow_custom_html: bool = Field(False, description="允许自定义 HTML")
    allow_view_password_history: bool = Field(False, description="允许查看历史密码")
    allow_batch_import: bool = Field(True, description="允许批量导入")
    allow_api_access: bool = Field(False, description="允许 API 访问")

    # 配额
    max_unlock_per_day: int = Field(100, ge=1, description="每日最大解锁次数")
    max_import_per_time: int = Field(100, ge=1, description="单次导入限制")

    # 价格
    price: Optional[Decimal] = Field(None, ge=0, description="价格")
    currency: str = Field("CNY", max_length=10, description="货币")

    sort_order: int = Field(0, description="排序")

    class Config:
        json_schema_extra = {
            "example": {
                "name": "专业版",
                "description": "适合中小企业",
                "max_accounts": 200,
                "max_share_pages": 20,
                "max_nodes": 5,
                "allow_custom_html": True,
                "price": 299.00
            }
        }


# ========== 更新套餐 ==========
class PackageUpdate(BaseModel):
    """更新套餐 Schema"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None)
    max_accounts: Optional[int] = Field(None, ge=1)
    max_share_pages: Optional[int] = Field(None, ge=1)
    max_nodes: Optional[int] = Field(None, ge=1)
    min_unlock_interval: Optional[int] = Field(None, ge=60)
    allow_custom_html: Optional[bool] = Field(None)
    allow_view_password_history: Optional[bool] = Field(None)
    allow_batch_import: Optional[bool] = Field(None)
    allow_api_access: Optional[bool] = Field(None)
    max_unlock_per_day: Optional[int] = Field(None, ge=1)
    max_import_per_time: Optional[int] = Field(None, ge=1)
    price: Optional[Decimal] = Field(None, ge=0)
    currency: Optional[str] = Field(None)
    is_active: Optional[bool] = Field(None)
    sort_order: Optional[int] = Field(None)


# ========== 套餐响应 ==========
class PackageResponse(BaseModel):
    """套餐响应 Schema"""
    id: int
    name: str
    description: Optional[str]
    max_accounts: int
    max_share_pages: int
    max_nodes: int
    min_unlock_interval: int
    allow_custom_html: bool
    allow_view_password_history: bool
    allow_batch_import: bool
    allow_api_access: bool
    max_unlock_per_day: int
    max_import_per_time: int
    price: Optional[Decimal]
    currency: str
    is_active: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/package.ts
import { get, post, put, del } from '@/utils/request'

/**
 * 套餐管理相关 API
 */

/**
 * 套餐类型
 */
export interface Package {
  id: number
  name: string
  description: string
  max_accounts: number
  max_share_pages: number
  max_nodes: number
  max_proxies: number
  max_concurrent_tasks: number
  min_unlock_interval: number
  allow_custom_html: boolean
  allow_view_password_history: boolean
  allow_batch_import: boolean
  allow_api_access: boolean
  max_unlock_per_day: number
  max_import_per_time: number
  price: number
  currency: string
  is_active: boolean
  sort_order: number
  created_at: string
  updated_at?: string
}

/**
 * 套餐列表参数（管理员）
 */
export interface PackageListParams {
  page?: number
  per_page?: number
  include_inactive?: boolean
}

/**
 * 套餐列表响应（管理员）
 */
export interface PackageListResponse {
  items: Package[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

/**
 * 创建套餐请求
 */
export interface CreatePackageRequest {
  name: string
  description: string
  max_accounts: number
  max_share_pages: number
  max_nodes: number
  max_proxies?: number
  max_concurrent_tasks?: number
  min_unlock_interval: number
  allow_custom_html: boolean
  allow_view_password_history: boolean
  allow_batch_import: boolean
  allow_api_access: boolean
  max_unlock_per_day: number
  max_import_per_time: number
  price: number
  currency?: string
  sort_order?: number
}

/**
 * 更新套餐请求
 */
export interface UpdatePackageRequest {
  name?: string
  description?: string
  max_accounts?: number
  max_share_pages?: number
  max_nodes?: number
  max_proxies?: number
  max_concurrent_tasks?: number
  min_unlock_interval?: number
  allow_custom_html?: boolean
  allow_view_password_history?: boolean
  allow_batch_import?: boolean
  allow_api_access?: boolean
  max_unlock_per_day?: number
  max_import_per_time?: number
  price?: number
  currency?: string
  is_active?: boolean
  sort_order?: number
}

/**
 * 获取套餐列表（公开接口）
 * 无需登录，所有用户可访问。仅返回激活的套餐。
 */
export function getPackageListApi() {
  return get<Package[]>('/packages')
}

/**
 * 管理员：获取套餐列表（含禁用）
 */
export function getAdminPackageListApi(params?: PackageListParams) {
  return get<PackageListResponse>('/admin/packages', params)
}

/**
 * 管理员：获取套餐详情
 */
export function getPackageDetailApi(id: number) {
  return get<Package>(`/admin/packages/${id}`)
}

/**
 * 管理员：创建套餐
 */
export function createPackageApi(data: CreatePackageRequest) {
  return post<Package>('/admin/packages', data)
}

/**
 * 管理员：更新套餐
 */
export function updatePackageApi(id: number, data: UpdatePackageRequest) {
  return put<Package>(`/admin/packages/${id}`, data)
}

/**
 * 管理员：删除套餐
 * 注意：删除套餐是永久性的，无法恢复
 * 如果有卡密或用户权限关联此套餐，建议禁用而不是删除
 */
export function deletePackageApi(id: number) {
  return del(`/admin/packages/${id}`)
}

/**
 * 管理员：启用套餐
 */
export function enablePackageApi(id: number) {
  return put(`/admin/packages/${id}`, { is_active: true })
}

/**
 * 管理员：禁用套餐
 */
export function disablePackageApi(id: number) {
  return put(`/admin/packages/${id}`, { is_active: false })
}
